### PR TITLE
update metadata & index tx signature

### PR DESCRIPTION
**Pull request by imabdulbasit**

### CODE REVIEW

File: api/src/mutations/drop.rs [hunk 1]
```diff
@@ -240,6 +243,157 @@ impl Mutation {
...
+        let AppContext { db, user_id, .. } = ctx.data::<AppContext>()?;
+        let UserID(id) = user_id;
+        let conn = db.get();
+        let producer = ctx.data::<Producer<NftEvents>>()?;
+        let nft_storage = ctx.data::<NftStorageClient>()?;
+        let solana = ctx.data::<Solana>()?;
+
+        let user_id = id.ok_or_else(|| Error::new("X-USER-ID header not found"))?;
+
+        let (drop_model, solana_collection_model) = Drops::find()
```
Comment (imabdulbasit) [inline api/src/mutations/drop.rs:283]: Have to change this to make it generic for all blockchains. Plan is to find drop and collection -> match the blockchain

File: api/src/entities/sea_orm_active_enums.rs [hunk 2]
```diff
@@ -19,8 +19,16 @@ pub enum Blockchain {
...
+    #[sea_orm(string_value = "canceled")]
+    Canceled,
     #[sea_orm(string_value = "created")]
     Created,
+    #[sea_orm(string_value = "failed")]
+    Failed,
     #[sea_orm(string_value = "pending")]
     Pending,
+    #[sea_orm(string_value = "rejected")]
+    Rejected,
```
Comment (kespinola) [inline api/src/entities/sea_orm_active_enums.rs:33]: Ok nice. can you incorporate into the drop status?

File: api/src/mutations/drop.rs [hunk 3]
```diff
@@ -240,6 +243,157 @@ impl Mutation {
             drop: Drop::new(drop_model, collection_model),
         })
     }
+
+    /// Res
+    ///
+    /// # Errors
+    /// This function fails if ...
```
Comment (kespinola) [inline api/src/mutations/drop.rs:250]:
```
/// This mutation allows updating a drop and it's associated collection by ID. It returns an error if it fails to reach the database, emit update events or assemble the on-chain transaction.
```

File: api/src/mutations/drop.rs (hunk 1 above)
Comment (kespinola) [reply api/src/mutations/drop.rs:283]: Should just do an extra query in the update method of the solana.editions().update() to find the required addresses and update the record.

File: api/src/mutations/drop.rs [hunk 4]
```diff
@@ -308,6 +503,35 @@ impl TryFrom<CollectionCreator> for Creator {
...
+#[derive(Debug, Clone, Serialize, Deserialize, InputObject)]
+pub struct EditDropInput {
+    pub drop: Uuid,
+    pub price: Option<u64>,
+    pub start_time: Option<DateTime<Utc>>,
+    pub end_time: Option<DateTime<Utc>>,
+    pub seller_fee_basis_points: Option<u16>,
+    pub metadata_json: Option<MetadataJsonInput>,
+}
+
```
Comment (kespinola) [inline api/src/mutations/drop.rs:534]:
```suggestion
/// Input object for editing a drop and associated collection by ID
#[derive(Debug, Clone, Serialize, Deserialize, InputObject)]
pub struct EditDropInput {
    /// The unique identifier of the drop
    pub drop: Uuid,
    /// The new price for the drop in the native token of the blockchain
    pub price: Option<u64>,
    /// The new start time for the drop in UTC
    pub start_time: Option<DateTime<Utc>>,
    /// The new end time for the drop in UTC
    pub end_time: Option<DateTime<Utc>>,
    /// The new seller fee basis points for the drop
    pub seller_fee_basis_points: Option<u16>,
    /// The new metadata JSON for the drop
    pub metadata_json: Option<MetadataJsonInput>,
}
```

**Review by kespinola — COMMENTED** [review]
This is looking good. Left some notes.

File: api/src/mutations/drop.rs [hunk 5]
```diff
@@ -241,6 +243,138 @@ impl Mutation {
...
+        } = input;
+
+        if price.is_none()
+            && start_time.is_none()
+            && end_time.is_none()
+            && seller_fee_basis_points.is_none()
+            && metadata_json.is_none()
+        {
+            return Err(Error::new("please select atleast one field to update"));
+        }
```
Comment (kespinola) [inline api/src/mutations/drop.rs:271]: Ok so this keeping all the previous settings? We should call this patch_drop then to signal to the caller they just need to alter what they want changed.

So input goes to PatchDropInput and PatchDropResponse.

Comment (kespinola) [reply api/src/mutations/drop.rs:271]:
```suggestion
        if vec![price, start_time, end_time, seller_fee_basis_points, metadata_json].iter().all(|i| i.is_none())
        {
            return Err(Error::new("please select atleast one field to update"));
        }
```

nb but an option

File: api/src/mutations/drop.rs [hunk 6]
```diff
@@ -241,6 +243,138 @@ impl Mutation {
...
+
+        let collection = collection_model.ok_or_else(|| Error::new("collection not found"))?;
+
+        let mut drop_am = drops::ActiveModel::from(drop_model.clone());
+
+        if let Some(price) = price {
+            drop_am.price = Set(price.try_into()?);
+        }
+        if let Some(start_time) = start_time {
+            drop_am.start_time = Set(Some(start_time.naive_utc()));
```
[the remark this answers is not in the document]
Comment (imabdulbasit) [reply api/src/mutations/drop.rs:298]: No, the sea-orm DateTime is actually NaiveDateTime and start_time is DateTime<Utc> so we are converting start_time with tz = Utc to NaiveDateTime

https://docs.rs/sea-orm/latest/src/sea_orm/entity/prelude.rs.html#28

File: api/src/mutations/drop.rs (hunk 5 above)
Comment (imabdulbasit) [reply api/src/mutations/drop.rs:271]: but all elements in vector should have same data type

Comment (imabdulbasit) [reply api/src/mutations/drop.rs:271]: > Ok so this keeping all the previous settings? We should call this patch_drop then to signal to the caller they just need to alter what they want changed.
> 
> So input goes to PatchDropInput and PatchDropResponse.

Do we want to send all the fields that can be updated as input? I thought we need to send the field which needs to be updated.

Comment (kespinola) [reply api/src/mutations/drop.rs:271]: Either. Edit would be a put operation which replaces all fields on the record making the caller send over all data that can change. Just sending what you want updated is a PATCH operation which is normally friendly so happy to go with this but should name the mutation patch_drop to make clear that the caller only needs to send the fields they want to update.

File: api/src/mutations/drop.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kespinola) [reply api/src/mutations/drop.rs:298]: Can we setup sea-orm to use chrono::DateTime<chrono::Utc>

https://docs.rs/sea-orm/latest/src/sea_orm/entity/prelude.rs.html#36

File: api/src/objects/drop.rs [hunk 7]
```diff
@@ -102,6 +102,17 @@ impl Drop {
...
+                _,
+                _,
+                _,
+                _,
+                CreationStatus::Blocked
+                | CreationStatus::Canceled
+                | CreationStatus::Failed
+                | CreationStatus::Rejected,
+            ) => Ok(DropStatus::Failed),
             (Some(true), ..) => Ok(DropStatus::Scheduled),
```
Comment (kespinola) [inline api/src/objects/drop.rs:116]: Thank you.

**Pull request merged**